File: nodejs/patterns/src/pattern_runner.js
const singleton = require("./creational/singleton");
const factoryMethod = require("./creational/factory_method");
const abstractFactory = require("./creational/abstract_factory");
const builder = require("./creational/builder");
const prototype = require("./creational/prototype");
const objectPool = require("./creational/object_pool");
const lazyInitialization = require("./creational/lazy_initialization");

const adapter = require("./structural/adapter");
const decorator = require("./structural/decorator");
const composite = require("./structural/composite");
const facade = require("./structural/facade");
const proxy = require("./structural/proxy");
const bridge = require("./structural/bridge");
const flyweight = require("./structural/flyweight");
const compositeEntity = require("./structural/composite_entity");

const observer = require("./behavioral/observer");
const strategy = require("./behavioral/strategy");
const command = require("./behavioral/command");
const state = require("./behavioral/state");
const templateMethod = require("./behavioral/template_method");
const chainOfResponsibility = require("./behavioral/chain_of_responsibility");
const iterator = require("./behavioral/iterator");
const mediator = require("./behavioral/mediator");
const memento = require("./behavioral/memento");

const PATTERNS = [
  { id: "singleton", category: "creational", demo: singleton.demo },
  { id: "factory-method", category: "creational", demo: factoryMethod.demo },
  { id: "abstract-factory", category: "creational", demo: abstractFactory.demo },
  { id: "builder", category: "creational", demo: builder.demo },
  { id: "prototype", category: "creational", demo: prototype.demo },
  { id: "object-pool", category: "creational", demo: objectPool.demo },
  { id: "lazy-initialization", category: "creational", demo: lazyInitialization.demo },
  { id: "adapter", category: "structural", demo: adapter.demo },
  { id: "decorator", category: "structural", demo: decorator.demo },
  { id: "composite", category: "structural", demo: composite.demo },
  { id: "facade", category: "structural", demo: facade.demo },
  { id: "proxy", category: "structural", demo: proxy.demo },
  { id: "bridge", category: "structural", demo: bridge.demo },
  { id: "flyweight", category: "structural", demo: flyweight.demo },
  { id: "composite-entity", category: "structural", demo: compositeEntity.demo },
  { id: "observer", category: "behavioral", demo: observer.demo },
  { id: "strategy", category: "behavioral", demo: strategy.demo },
  { id: "command", category: "behavioral", demo: command.demo },
  { id: "state", category: "behavioral", demo: state.demo },
  { id: "template-method", category: "behavioral", demo: templateMethod.demo },
  { id: "chain-of-responsibility", category: "behavioral", demo: chainOfResponsibility.demo },
  { id: "iterator", category: "behavioral", demo: iterator.demo },
  { id: "mediator", category: "behavioral", demo: mediator.demo },
  { id: "memento", category: "behavioral", demo: memento.demo },
];

const patternById = new Map(PATTERNS.map((pattern) => [pattern.id, pattern]));

function normalizePatternId(patternId) {
  return String(patternId).trim().toLowerCase().replace(/[_\s]+/g, "-");
}

function listPatterns() {
  return PATTERNS.map(({ id, category }) => ({ id, category }));
}

function runPattern(patternId) {
  const normalized = normalizePatternId(patternId);
  const pattern = patternById.get(normalized);
  if (!pattern) {
    throw new Error(`Unknown pattern: ${patternId}`);
  }
  return pattern.demo();
}

function runAllPatterns() {
  const results = {};
  for (const pattern of PATTERNS) {
    results[pattern.id] = pattern.demo();
  }
  return results;
}

module.exports = {
  normalizePatternId,
  listPatterns,
  runPattern,
  runAllPatterns,
};

File: python/patterns/src/prototype_patterns/pattern_runner.py
"""Pattern registry and helpers for running demos one at a time."""

from __future__ import annotations

import json
from dataclasses import asdict, is_dataclass
from typing import Callable

from prototype_patterns.behavioral import (
    chain_of_responsibility,
    command,
    iterator,
    mediator,
    memento,
    observer,
    state,
    strategy,
    template_method,
)
from prototype_patterns.creational import (
    abstract_factory,
    builder,
    factory_method,
    lazy_initialization,
    object_pool,
    prototype,
    singleton,
)
from prototype_patterns.structural import (
    adapter,
    bridge,
    composite,
    composite_entity,
    decorator,
    facade,
    flyweight,
    proxy,
)

PatternDemo = Callable[[], object]
PatternSpec = tuple[str, str, PatternDemo]

PATTERNS: tuple[PatternSpec, ...] = (
    ("singleton", "creational", singleton.demo),
    ("factory-method", "creational", factory_method.demo),
    ("abstract-factory", "creational", abstract_factory.demo),
    ("builder", "creational", builder.demo),
    ("prototype", "creational", prototype.demo),
    ("object-pool", "creational", object_pool.demo),
    ("lazy-initialization", "creational", lazy_initialization.demo),
    ("adapter", "structural", adapter.demo),
    ("decorator", "structural", decorator.demo),
    ("composite", "structural", composite.demo),
    ("facade", "structural", facade.demo),
    ("proxy", "structural", proxy.demo),
    ("bridge", "structural", bridge.demo),
    ("flyweight", "structural", flyweight.demo),
    ("composite-entity", "structural", composite_entity.demo),
    ("observer", "behavioral", observer.demo),
    ("strategy", "behavioral", strategy.demo),
    ("command", "behavioral", command.demo),
    ("state", "behavioral", state.demo),
    ("template-method", "behavioral", template_method.demo),
    ("chain-of-responsibility", "behavioral", chain_of_responsibility.demo),
    ("iterator", "behavioral", iterator.demo),
    ("mediator", "behavioral", mediator.demo),
    ("memento", "behavioral", memento.demo),
)

_PATTERN_BY_ID: dict[str, tuple[str, PatternDemo]] = {
    pattern_id: (category, demo) for pattern_id, category, demo in PATTERNS
}


def normalize_pattern_id(pattern_id: str) -> str:
    """Normalize input so users can pass spaces/underscores/hyphens."""
    return "-".join(pattern_id.strip().lower().replace("_", " ").split())


def list_patterns() -> list[tuple[str, str]]:
    """Return all patterns in deterministic category/id order."""
    return [(pattern_id, category) for pattern_id, category, _ in PATTERNS]


def run_pattern(pattern_id: str) -> object:
    """Run exactly one pattern by id."""
    normalized = normalize_pattern_id(pattern_id)
    spec = _PATTERN_BY_ID.get(normalized)
    if spec is None:
        raise ValueError(f"Unknown pattern: {pattern_id}")
    _, demo = spec
    return demo()


def run_all_patterns() -> dict[str, object]:
    """Run all demos, preserving pattern order."""
    return {pattern_id: demo() for pattern_id, _, demo in PATTERNS}


def format_result(result: object) -> str:
    """Serialize demo output for CLI display."""
    if is_dataclass(result):
        result = asdict(result)
    elif isinstance(result, tuple):
        result = list(result)
    return json.dumps(result)
